### PR TITLE
pkgutil: add update all, check-mode, squashing and examples

### PR DESCRIPTION
**Pull request by mavit**
Taken from https://github.com/ansible/ansible/pull/51651 by @dagwieers, which was taken from https://github.com/ansible/ansible/pull/27866 by @scathatheworm.  Let’s have one last attempt to get this merged.

> ##### SUMMARY
>
> Original PR #27866 from @scathatheworm
>
> When working with Solaris pkgutil CSW packages, I came across this module
being very basic in functionality, in particular, that I could not use it
to update all CSW packages.
>
> When going into details into the code I also found it did not incorporate
a possibility of doing dry-run from the underlying utility, or supported to
specify multiple packages for operations.
>
> This module probably sees very little use, but it seemed like nice
functionality to add and make it behave a little more like other package
modules.
> ##### ISSUE TYPE
>
>     * Feature Pull Request
>
>
> ##### COMPONENT NAME
>
> pkgutil module
> ##### ANSIBLE VERSION
>
> ```
> ansible 2.3.1.0
>   config file = /etc/ansible/ansible.cfg
>   configured module search path = Default w/o overrides
>   python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5
20150623 (Red Hat 4.8.5-4)]
> ```
>
> ##### ADDITIONAL INFORMATION
>
>     * Added ability to upgrade all packages:
>
>
> ```yaml
> - pkgutil:
>     name: '*'
>     state: latest
> ```
>
>     * Added ability to modify state of a list of packages:
>
>
> ```yaml
> - pkgutil:
>     name:
>     - CSWtop
>     - CSWwget
>     - CSWlsof
>     state: present
> ```
>
>     * Added ability to have underlying tool perform a dry-run when using
check mode, pkgutil -n
>
>     * Added ability to configure force option to force packages to state
determined by repository (downgrade for example)
>
>
> ```yaml
> - pkgutil:
>     name: CSWtop
>     state: latest
>     force: yes
> ```
>
>     * Added more examples and documentation to show the new functionality

### CODE REVIEW

File: changelogs/fragments/pkgutil-check-mode-etc.yaml
Comment (felixfontein) [inline changelogs/fragments/pkgutil-check-mode-etc.yaml:4]:
```suggestion
  - pkgutil - module now supports check mode (https://github.com/ansible-collections/community.general/pull/799).
  - pkgutil - module can now accept a list of packages (https://github.com/ansible-collections/community.general/pull/799).
  - pkgutil - module has a new option, ``force``, equivalent to the ``-f`` option to the `pkgutil <http://pkgutil.net/>`_ command (https://github.com/ansible-collections/community.general/pull/799).
```

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:23]:
```suggestion
```

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:34]:
```suggestion
    elements: str
```

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:193]:
```suggestion
            name=dict(type='list', elements='str', required=True, aliases=['pkg']),
```
Why do you add an alias here?

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:194]: Why add more choices? Are they needed?

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:55]:
```suggestion
    - Allows the update process to downgrade packages to what is present in the repository set this to C(yes).
```
I'm not really sure what this sentence means.

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:50]:
```suggestion
    - If you want to refresh your catalog from the mirror, set this to C(yes).
```

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:57]:
```suggestion
    type: bool
    version_added: 1.2.0
```

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:192]: I'm not the original author of this, but a `pkg` alias seems reasonable to me for consistency with, e.g., the `yum` module.

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:193]: Again, this seems to be for consistency with, e.g., the `yum` module.

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:46]: You should also document `installed` and `removed` (or remove them).

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:47]: I think it is now discouraged to have a default for `state`. In any case, since it hasn't been here before, it won't hurt to not add it (yet).

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:128]: Should this be specified if check mode is enabled? I don't know pkgutil, but I guess this actually does modify something.

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:158]: If both `-n` and `-U` are specified, does it modify the state somehow?

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:207]: Maybe do a check for `'*' in name and name != ['*']`, and fail in that case. I don't think pkgutil likes to get `*` passed (without knowing it).

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:158]: No.  We can see from the `pkgutil` [source code ](https://sourceforge.net/p/pkgutil/code/HEAD/tree/trunk/pkgutil) that it exits just before fetching the catalog if `-n` is specified.  This isn't directly affected by `-U`.

Note however that the catalog may get updated anyway even in check mode if `state` is `latest`, because that may happen as a part of checking for the latest.  I think this is reasonable and expected.

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:128]: I think it is reasonable and expected that the catalog may be updated as a part of checking for new packages.  It's just a cached copy of the list of packages available in the repository.

`-U` means "always refresh the catalog, even if it was refreshed recently".  Even without `-U`, it's quite likely that the catalog will be refreshed.

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:207]: It says, `Package * not in catalog. Exiting.`, returning 1, which I don't think is any worse than detecting this ourselves.

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:46]: I think it's already clear that `installed` is for installing and `removed` is for removing, but sure.

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [reply plugins/modules/packaging/os/pkgutil.py:207]: Except if in the future `pkgutil` supports `*` :)

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [reply plugins/modules/packaging/os/pkgutil.py:128]: The documentation about check mode is pretty explicit: https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html#validating-tasks-check-mode-and-diff-mode

> In check mode, Ansible runs **without making any changes** on remote systems.

(emphasis mine). Modules that implement check mode have to make sure they do not modify the system when run in check mode. If I would be a user of this module, I would consider such a behavior a serious bug.

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:207]: The [most-recent commit to `pkgutil`](https://sourceforge.net/p/pkgutil/code/commit_browser) was in 2014, so don't hold your breath waiting for this (or any) new feature!

File: plugins/modules/packaging/os/pkgutil.py
Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:128]: Some writes to disk are clearly permitted, since Ansible cannot run without creating temporary files or generating logs on the remote system.

The Ansible `yum` module exhibits the same behaviour, so I think people must consider this to be acceptable.

Comment (felixfontein) [reply plugins/modules/packaging/os/pkgutil.py:128]: If you really want to do this, you should at least document this behavior.

Comment (mavit) [reply plugins/modules/packaging/os/pkgutil.py:128]: I’m not sure it’s helpful to obfuscate the documentation with implementation details that users aren’t going to care about or be surprised by, but okay.

File: plugins/modules/packaging/os/pkgutil.py
Comment (felixfontein) [inline plugins/modules/packaging/os/pkgutil.py:20]:
```suggestion
- Unlike the M(community.general.svr4pkg) module, it will resolve and download dependencies.
```

**Comment by felixfontein** [comment]
@dermute @dagwieers @scathatheworm could you please take a look at this PR? I'd like to have another opinion of someone actually using this module :)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
If nobody complains until tomorrow morning, I'll merge and backport this so it gets included in community.general 1.2.0.

**Comment by felixfontein** [comment]
@mavit thanks a lot for reviving this!

**Comment by mavit** [comment]
@felixfontein Thanks for the fast and thorough review!

**Pull request merged**
